test/e2e/resource_usage_gatherer.go

```diff
@@ -75,43 +70,193 @@ func (s *ResourceUsageSummary) PrintJSON() string {
+	c                    *client.Client
+	nodeName             string
+	wg                   *sync.WaitGroup
+	containerIDToNameMap map[string]string
+	containerIDs         []string
+	stopCh               chan struct{}
+	dataSeries           []resourceUsagePerContainer
+}
+
+func (w *resourceGatherWorker) singleProbe(errors []error) {
+	data := make(resourceUsagePerContainer)
+	nodeUsage, err := getOneTimeResourceUsageOnNode(w.c, w.nodeName, 15*time.Second, func() []string { return w.containerIDs }, true)
+	if err != nil {
+		errors = append(errors, err)
+		return
+	}
+	for k, v := range nodeUsage {
+		data[w.containerIDToNameMap[k]] = v
+	}
+	w.dataSeries = append(w.dataSeries, data)
+}
+
+func (w *resourceGatherWorker) gather(initialSleep time.Duration, errors []error) {
+	defer utilruntime.HandleCrash()
+	defer w.wg.Done()
+	select {
+	case <-time.After(initialSleep):
+		w.singleProbe(errors)
 		for {
 			select {
-			case <-time.After(period):
-				now := time.Now()
-				data, err := g.getKubeSystemContainersResourceUsage(c)
-				if err != nil {
-					Logf("Error while getting resource usage: %v", err)
-					continue
-				}
-				g.usageTimeseries[now] = data
-			case <-g.stopCh:
-				Logf("Stop channel is closed. Stopping gatherer.")
-				return nil
+			case <-time.After(resourceDataGatheringPeriod):
+				w.singleProbe(errors)
+			case <-w.stopCh:
+				return
 			}
 		}
+	case <-w.stopCh:
+		return
+	}
+}
+
+func (g *containerResourceGatherer) getKubeSystemContainersResourceUsage(c *client.Client) error {
+	errors := make([]error, 0)
+	delay := resourceDataGatheringPeriod / time.Duration(len(g.workers))
+	go func() {
+		for i := range g.workers {
+			go g.workers[i].gather(delay, errors)
+		}
 	}()
+	g.workerWg.Wait()
+
+	if len(errors) != 0 {
+		return fmt.Errorf("Errors while gathering usage data: %v", errors)
+	}
+	return nil
+}
+
+type containerResourceGatherer struct {
+	client               *client.Client
+	stopCh               chan struct{}
+	workers              []resourceGatherWorker
+	workerWg             sync.WaitGroup
+	containerIDToNameMap map[string]string
+	containerIDs         []string
+}
+
+func NewResourceUsageGatherer(c *client.Client) *containerResourceGatherer {
+	g := containerResourceGatherer{
+		client:               c,
+		stopCh:               make(chan struct{}),
+		containerIDToNameMap: make(map[string]string),
+		containerIDs:         make([]string, 0),
+	}
+
+	pods, err := c.Pods("kube-system").List(api.ListOptions{})
+	if err != nil {
+		Logf("Error while listing Pods: %v", err)
+		return nil
+	}
+	for _, pod := range pods.Items {
+		for _, container := range pod.Status.ContainerStatuses {
+			containerID := strings.TrimPrefix(container.ContainerID, "docker:/")
+			g.containerIDToNameMap[containerID] = pod.Name + "/" + container.Name
+			g.containerIDs = append(g.containerIDs, containerID)
+		}
+	}
+	var nodeList *api.NodeList
+	if wait.PollImmediate(poll, 10*time.Second, func() (bool, error) {
```

Comment:
Did you see this being a problem? Why not simply nodes, err L- c.Nodes().List(api.ListOptions{}) ?